<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to keju • keju</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to keju">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">keju</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction.html">Introduction to keju</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pimentellab/keju/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to keju</h1>
                        <h4 data-toc-skip class="author">Albert Xue</h4>
            
            <h4 data-toc-skip class="date">February 17 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pimentellab/keju/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>keju is a set of models for statistical inference in Massively
Parallel Reporter Assay (MPRA) data. It takes in raw counts to assign
each enhancer a transcription rate and, optionally, an effect size that
denotes the difference between transcription rates in a control
treatment and an alternate treatment.</p>
<p>As a package, keju will filter and process your data for you before
fitting the model. Alternatively, you can provide pre-filtered counts
and skip the keju filtering step. Please just note that you must provide
raw unnormalized counts (or, at the minimum, integers).</p>
<p>As a note on terminology, “architecture” can be a little bit
confusing. To keju, an “architecture” is something to infer a
transcription rate and/or differential activity for. So an
“architecture” can be a ref/alt SNP pair (in which case the “control
condition” would be the reference allele and the “alternate treatment”
would be the alternate allele), a regulatory element in different
conditions (think drug effect comparison), etc.</p>
<div class="section level3">
<h3 id="choosing-the-correct-keju-model">Choosing the correct keju model<a class="anchor" aria-label="anchor" href="#choosing-the-correct-keju-model"></a>
</h3>
<p>Choosing the correct model can be somewhat confusing. keju is not a
singular model, but a suite of models for different use cases that fit
within each other, a little like Russian nesting dolls.</p>
<p>To summarize the next few paragraphs,</p>
<ul>
<li>the simplest modeling option is <code>no_motif</code>, and that will
get you most of the benefits of using keju. If you’re not sure which
model to use, you should probably at least start with
<code>no_motif</code>.</li>
<li>
<code>motif_shrinkage</code> is slightly more specialized version of
<code>no_motif</code>. If architectures have some kind of shared
structure (i.e., multiple architectures test the same transcription
factor binding motif), <code>motif_shrinkage</code> has some statistical
niceties that may be of interest and slightly better performance given
this motif-level metadata.</li>
<li>
<code>covariate_motif_slope_intercept</code> is a slightly more
specialized version of <code>motif_shrinkage</code>. If you think your
covariates are affecting the transcription rate of your architectures
(i.e. minimal promoter choice (see our paper)),
<code>covariate_motif_slope_intercept</code> can quantify those effects
for you given motif-level and covariate-level metadata.</li>
</ul>
<p>The most general model, and our default recommended starting point,
is <code>no_motif</code>. <code>no_motif</code> makes no assumptions on
correlations between enhancers or architectures, and shrinks
transcription rate estimates and effect size estimates towards a generic
standard normal prior. Use <code>no_motif</code> if you don’t have a
concrete structure among your tested architectures, which many users
will not. Even if you do have some kind of structure,
<code>no_motif</code> will give you viable estimates, just without some
motif-level bells and whistles.</p>
<p>In contrast, some users can use <code>motif_shrinkage</code> if they
have an exploitable structure in their tested architectures. An example
of motif-level structure in the architectures is provided below from the
<a href="https://www.nature.com/articles/s41467-024-54502-9" class="external-link">Zahm et
al.</a> data, where several architectures are all actually testing the
same Rarb transcription factor binding motif. In this case,
<code>motif_shrinkage</code> shrinks estimates towards a shared
motif-level mean, and will also provide <em>motif-level</em> estimates
for transcription rate and effect sizes, not just architecture-level
estimates. For example, in the example below <code>no_motif</code> and
<code>motif_shrinkage</code> will both provide 18 transcription rate
estimates and 18 effect size estimates, one for each architecture.
However, <code>motif_shrinkage</code> will also provide a transcription
rate estimate and effect size estimate for the motif itself, and the
architecture-level estimates will be regularized to these motif-level
estimates. Use <code>motif_shrinkage</code> if you have some kind of
structure among your architectures. If you do not know if your structure
qualifies, we recommend just using <code>no_motif</code>.</p>
<p>
<img src="../reference/figures/motif.png" width="300"></p>
<p>Finally, the most specialized model is
<code>covariate_motif_slope_intercept</code>, also the full
<strong>keju</strong> model.
<code>covariate_motif_slope_intercept</code> requires motif-level
structure, and is used in the case where the covariates provided in the
model have interesting effects on transcription rate that we want to
disentangle. As an example, in the <a href="https://www.nature.com/articles/s41467-024-54502-9" class="external-link">Zahm et
al.</a> data, choice of minimal promoter strongly affects transcription
rate estimates of each architecture (see Figure 4 in our paper). Use
<code>covariate_motif_slope_intercept</code> if you want covariate-level
slope and intercept estimates on transcription rate.</p>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="loading-the-input-data">Loading the input data<a class="anchor" aria-label="anchor" href="#loading-the-input-data"></a>
</h3>
<p>We load the input data here, which contains the first 100
architectures (in alphabetical order) of SF_DEX in our paper. This data
also contains a maximum of 20 barcodes per architecture for runtime’s
sake.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pimentellab/keju" class="external-link">keju</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.4.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:keju':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sf_dex.rda"</span>, package<span class="op">=</span><span class="st">"keju"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt;        X Unnamed..0 RNA_sample_number treatment DNA_sample_number</span></span>
<span><span class="co">#&gt; 1  11149      11149                 1        SF                42</span></span>
<span><span class="co">#&gt; 2  19492      19492                 1        SF                42</span></span>
<span><span class="co">#&gt; 3 107905     107905                 1        SF                42</span></span>
<span><span class="co">#&gt; 4 158325     158325                 1        SF                42</span></span>
<span><span class="co">#&gt; 5 176049     176049                 1        SF                42</span></span>
<span><span class="co">#&gt; 6 207146     207146                 1        SF                42</span></span>
<span><span class="co">#&gt;                    barcode DNA_count  DNA_rpm                 architecture</span></span>
<span><span class="co">#&gt; 1 GCTTCCTTGTTAGTTATTAAGATA       392 4.413638 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt; 2 AAGTGTTTGCGTCAACCTTACAAA       341 3.839415 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt; 3 CCTTCCCACATACTACGACATGCA       182 2.049189 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt; 4 GATTCTGACTGGGATCTGAATTCC       146 1.643855 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt; 5 ATTGATTAAATATAAAGGTATACC       135 1.520003 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt; 6 CTTAGTTAGTTTCTCGGGACTTTG       118 1.328595 ALX3-1:1223, 1, set2, minCMV</span></span>
<span><span class="co">#&gt;       class barcode_n RNA_count   RNA_rpm</span></span>
<span><span class="co">#&gt; 1 candidate         1        39 1.6640826</span></span>
<span><span class="co">#&gt; 2 candidate         2        94 4.0108659</span></span>
<span><span class="co">#&gt; 3 candidate         3        34 1.4507387</span></span>
<span><span class="co">#&gt; 4 candidate         4         7 0.2986815</span></span>
<span><span class="co">#&gt; 5 candidate         5         5 0.2133439</span></span>
<span><span class="co">#&gt; 6 candidate         6        39 1.6640826</span></span></code></pre></div>
<p>Here, we mark the control treatment (SF, denoting “Serum Free”) and
the negative control architectures. keju uses negative controls to
correct for global shifts in differential activity, and can do so in a
covariate-specific manner. We also provide the covariates we want to
correct effect sizes with (in this case, it’s the minimal promoter used
in the architecture).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">$</span><span class="va">is_control_treatment</span> <span class="op">=</span> <span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">treatment</span> <span class="op">==</span> <span class="st">'SF'</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">is_control_architecture</span> <span class="op">=</span> <span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">class</span> <span class="op">==</span> <span class="st">'scramble'</span> <span class="op">|</span> <span class="va">df</span><span class="op">$</span><span class="va">class</span> <span class="op">==</span> <span class="st">'spacer'</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">covariates</span> <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split_fixed</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">architecture</span>, <span class="st">', '</span>, <span class="fl">4</span><span class="op">)</span><span class="op">[</span>, <span class="fl">4</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="no_motif">no_motif<a class="anchor" aria-label="anchor" href="#no_motif"></a>
</h3>
<p>This section details running keju in the no_motif setting. For
motif_shrinkage or covariate_motif_slope_intercept, there are other
sections below. Note that the processing step also changes between
different models.</p>
<p>Here, we build the keju object from our data. In our case, we have a
control treatment (SF) and an alternate treatment (Dex), and are
interested in differential activity between the two (hence,
infer_differential_activity=TRUE). If you only have a single treatment,
and are only interested in baseline transcription rate in that single
treatment, simply set infer_differential_activity = FALSE.</p>
<p>Note that the DNA and RNA counts are matched by (barcode, RNA batch),
and every argument in keju_from_counts is a length N vector (or castable
to a length N vector), where N is the number of observations. In our
experiment, we have 100 architectures, where each architecture has a
maximum of 20 barcodes, and has observations in two RNA batches. This
comes out to be 8230 observations, where each observation has the same
barcode and RNA batch, but not necessarily the same DNA batch or DNA
counts (but they can be the same!).</p>
<p>If this is the first time you have run keju, basilisk will install a
conda environment with the necessary prerequisites for you.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build the keju object</span></span>
<span><span class="va">keju</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_from_counts.html">keju_from_counts</a></span><span class="op">(</span></span>
<span>    barcode<span class="op">=</span><span class="va">df</span><span class="op">$</span><span class="va">barcode</span>,</span>
<span>    R <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_count</span>,</span>
<span>    D <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_count</span>,</span>
<span>    batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_sample_number</span>,</span>
<span>    dna_batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_sample_number</span>,</span>
<span>    architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">architecture</span>,</span>
<span>    treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">treatment</span>,</span>
<span>    is_control_treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_treatment</span>,</span>
<span>    is_control_architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_architecture</span>,</span>
<span>    covariates<span class="op">=</span><span class="va">df</span><span class="op">$</span><span class="va">covariates</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># optional filtering step</span></span>
<span><span class="va">keju</span> <span class="op">&lt;-</span> <span class="fu">keju</span><span class="fu">::</span><span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">keju</span><span class="op">)</span></span>
<span><span class="co">#&gt; Converting 425 NA values to zero</span></span>
<span><span class="co">#&gt; Filtering 30 architectures with insufficient barcodes.</span></span>
<span></span>
<span><span class="co"># process the filtered data to fit no_motif</span></span>
<span><span class="va">keju</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/use_no_motif.html">use_no_motif</a></span><span class="op">(</span><span class="va">keju</span>, infer_differential_activity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here we set the output directory for our model outputs. keju will
automatically create this directory if it finds it missing. Nicely
processed estimates will be output here.</p>
<p>Since we are interested in the no_motif model in this subsection, we
set model=‘no_motif’. Fitting this toy dataset only takes a few minutes,
but it can take hours on a full size dataset. Generally speaking, keju
also requires the use of four cores, one per fit chain. Otherwise it
will fit the chains serially (see cmdstanr documentation for more).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="st">'keju_vignette'</span></span>
<span></span>
<span><span class="co"># fit no_motif</span></span>
<span><span class="va">keju</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_fit.html">keju_fit</a></span><span class="op">(</span><span class="va">keju</span>, output_dir<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">'/no_motif'</span><span class="op">)</span>, model<span class="op">=</span><span class="st">'no_motif'</span>, infer_differential_activity<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running MCMC with 4 parallel chains...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 finished in 93.5 seconds.</span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 finished in 145.2 seconds.</span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 finished in 163.9 seconds.</span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 finished in 181.7 seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All 4 chains finished successfully.</span></span>
<span><span class="co">#&gt; Mean chain execution time: 146.1 seconds.</span></span>
<span><span class="co">#&gt; Total execution time: 181.7 seconds.</span></span>
<span><span class="co">#&gt; Warning: 8 of 4000 (0.0%) transitions hit the maximum treedepth limit of 10.</span></span>
<span><span class="co">#&gt; See https://mc-stan.org/misc/warnings for details.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Warning: 8 of 4000 (0.0%) transitions hit the maximum treedepth limit of 10.</span></span>
<span><span class="co">#&gt; See https://mc-stan.org/misc/warnings for details.</span></span>
<span><span class="co">#&gt; $num_divergent</span></span>
<span><span class="co">#&gt; [1] 0 0 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_max_treedepth</span></span>
<span><span class="co">#&gt; [1] 5 0 0 3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ebfmi</span></span>
<span><span class="co">#&gt; [1] 0.9536814 1.0435768 0.9728513 1.0570126</span></span>
<span><span class="co">#&gt; Warning: Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="motif_shrinkage">motif_shrinkage<a class="anchor" aria-label="anchor" href="#motif_shrinkage"></a>
</h3>
<p>This section details usage of the model motif_shrinkage. Note that
both the preprocessing and fitting steps change slightly.</p>
<p>Because we are using motif-level information, we need to provide the
appropriate metadata for preprocessing. Note that here, we are providing
the motif-level information in keju_from_counts.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">$</span><span class="va">motif</span> <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split_fixed</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">architecture</span>, <span class="st">':'</span>, <span class="fl">2</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="va">keju_ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_from_counts.html">keju_from_counts</a></span><span class="op">(</span></span>
<span>    barcode<span class="op">=</span><span class="va">df</span><span class="op">$</span><span class="va">barcode</span>,</span>
<span>    R <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_count</span>,</span>
<span>    D <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_count</span>,</span>
<span>    batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_sample_number</span>,</span>
<span>    dna_batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_sample_number</span>,</span>
<span>    architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">architecture</span>,</span>
<span>    treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">treatment</span>,</span>
<span>    is_control_treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_treatment</span>,</span>
<span>    is_control_architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_architecture</span>,</span>
<span>    covariates <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">covariates</span>,</span>
<span>    motif <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">motif</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We still need to filter the counts, but the processing function we
call is different (use_motif_shrinkage instead of use_no_motif).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keju_ms</span> <span class="op">&lt;-</span> <span class="fu">keju</span><span class="fu">::</span><span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">keju_ms</span><span class="op">)</span></span>
<span><span class="co">#&gt; Converting 425 NA values to zero</span></span>
<span><span class="co">#&gt; Filtering 30 architectures with insufficient barcodes.</span></span>
<span><span class="va">keju_ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/use_motif_shrinkage.html">use_motif_shrinkage</a></span><span class="op">(</span><span class="va">keju_ms</span>, infer_differential_activity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we specify motif_shrinkage through the model argument.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keju_ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_fit.html">keju_fit</a></span><span class="op">(</span><span class="va">keju_ms</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">'/motif_shrinkage'</span><span class="op">)</span>, model<span class="op">=</span><span class="st">'motif_shrinkage'</span>, infer_differential_activity<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running MCMC with 4 parallel chains...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 finished in 99.2 seconds.</span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 finished in 105.7 seconds.</span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 finished in 108.2 seconds.</span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 finished in 109.3 seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All 4 chains finished successfully.</span></span>
<span><span class="co">#&gt; Mean chain execution time: 105.6 seconds.</span></span>
<span><span class="co">#&gt; Total execution time: 109.4 seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_divergent</span></span>
<span><span class="co">#&gt; [1] 0 0 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_max_treedepth</span></span>
<span><span class="co">#&gt; [1] 0 0 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ebfmi</span></span>
<span><span class="co">#&gt; [1] 0.9976810 1.0551858 0.9785017 1.0254429</span></span>
<span><span class="co">#&gt; Warning: Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="covariate_motif_slope_intercept">covariate_motif_slope_intercept<a class="anchor" aria-label="anchor" href="#covariate_motif_slope_intercept"></a>
</h3>
<p>This section details usage of the model
covariate_motif_slope_intercept. We against start with creating the keju
object and filtering counts.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keju_cmsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_from_counts.html">keju_from_counts</a></span><span class="op">(</span></span>
<span>    barcode<span class="op">=</span><span class="va">df</span><span class="op">$</span><span class="va">barcode</span>,</span>
<span>    R <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_count</span>,</span>
<span>    D <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_count</span>,</span>
<span>    batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">RNA_sample_number</span>,</span>
<span>    dna_batch <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">DNA_sample_number</span>,</span>
<span>    architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">architecture</span>,</span>
<span>    treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">treatment</span>,</span>
<span>    is_control_treatment <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_treatment</span>,</span>
<span>    is_control_architecture <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">is_control_architecture</span>,</span>
<span>    covariates <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">covariates</span>,</span>
<span>    motif <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">motif</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">keju_cmsi</span> <span class="op">&lt;-</span> <span class="fu">keju</span><span class="fu">::</span><span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">keju_cmsi</span><span class="op">)</span></span>
<span><span class="co">#&gt; Converting 425 NA values to zero</span></span>
<span><span class="co">#&gt; Filtering 30 architectures with insufficient barcodes.</span></span></code></pre></div>
<p>Maybe predictably, the preprocessing function we use is
use_covariate_slope_intercept instead of use_no_motif or
use_motif_shrinkage.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keju_cmsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/use_covariate_slope_intercept.html">use_covariate_slope_intercept</a></span><span class="op">(</span><span class="va">keju_cmsi</span>, infer_differential_activity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We then fit the model, again specifying
covariate_motif_slope_intercept through the model argument.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keju_cmsi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keju_fit.html">keju_fit</a></span><span class="op">(</span><span class="va">keju_cmsi</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">'/covariate_motif_slope_intercept'</span><span class="op">)</span>, </span>
<span>                 model<span class="op">=</span><span class="st">'covariate_motif_slope_intercept'</span>, </span>
<span>                 infer_differential_activity<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running MCMC with 4 parallel chains...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:    1 / 2000 [  0%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  100 / 2000 [  5%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  200 / 2000 [ 10%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  300 / 2000 [ 15%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  400 / 2000 [ 20%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  500 / 2000 [ 25%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  600 / 2000 [ 30%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  700 / 2000 [ 35%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  800 / 2000 [ 40%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration:  900 / 2000 [ 45%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1000 / 2000 [ 50%]  (Warmup) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1001 / 2000 [ 50%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1100 / 2000 [ 55%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1200 / 2000 [ 60%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1300 / 2000 [ 65%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1400 / 2000 [ 70%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1500 / 2000 [ 75%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1600 / 2000 [ 80%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1700 / 2000 [ 85%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1800 / 2000 [ 90%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 1 finished in 101.5 seconds.</span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 1900 / 2000 [ 95%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 3 finished in 104.1 seconds.</span></span>
<span><span class="co">#&gt; Chain 2 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 2 finished in 105.6 seconds.</span></span>
<span><span class="co">#&gt; Chain 4 Iteration: 2000 / 2000 [100%]  (Sampling) </span></span>
<span><span class="co">#&gt; Chain 4 finished in 107.6 seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All 4 chains finished successfully.</span></span>
<span><span class="co">#&gt; Mean chain execution time: 104.7 seconds.</span></span>
<span><span class="co">#&gt; Total execution time: 107.8 seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_divergent</span></span>
<span><span class="co">#&gt; [1] 0 0 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_max_treedepth</span></span>
<span><span class="co">#&gt; [1] 0 0 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ebfmi</span></span>
<span><span class="co">#&gt; [1] 1.0016497 0.9916731 0.9401170 0.9802777</span></span>
<span><span class="co">#&gt; Warning: Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span>
<span><span class="co">#&gt; Setting row names on a tibble is deprecated.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="accessing-estimates">Accessing estimates<a class="anchor" aria-label="anchor" href="#accessing-estimates"></a>
</h3>
<p>keju provides both transcription rate estimates (alpha) and effect
size/differential activity estimates (beta). We can access both
estimates as shown below. The most important columns here are mean (the
mean estimate), sd (standard deviation of estimate), and
is_significant.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># alpha estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju</span><span class="op">$</span><span class="va">alphas_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 11</span></span></span>
<span><span class="co">#&gt;   variable   mean median     sd    mad     q5     q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> alpha[1] -<span style="color: #BB0000;">0.383</span> -<span style="color: #BB0000;">0.385</span> 0.088<span style="text-decoration: underline;">4</span> 0.087<span style="text-decoration: underline;">7</span> -<span style="color: #BB0000;">0.524</span> -<span style="color: #BB0000;">0.236</span>   1.00    <span style="text-decoration: underline;">5</span>118.    <span style="text-decoration: underline;">2</span>706.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> alpha[2] -<span style="color: #BB0000;">0.177</span> -<span style="color: #BB0000;">0.178</span> 0.082<span style="text-decoration: underline;">9</span> 0.084<span style="text-decoration: underline;">6</span> -<span style="color: #BB0000;">0.315</span> -<span style="color: #BB0000;">0.041</span><span style="color: #BB0000; text-decoration: underline;">3</span>  1.00    <span style="text-decoration: underline;">5</span>032.    <span style="text-decoration: underline;">3</span>009.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> alpha[3] -<span style="color: #BB0000;">0.913</span> -<span style="color: #BB0000;">0.913</span> 0.085<span style="text-decoration: underline;">0</span> 0.085<span style="text-decoration: underline;">5</span> -<span style="color: #BB0000;">1.05</span>  -<span style="color: #BB0000;">0.772</span>   1.00    <span style="text-decoration: underline;">4</span>492.    <span style="text-decoration: underline;">2</span>971.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> alpha[4] -<span style="color: #BB0000;">0.658</span> -<span style="color: #BB0000;">0.657</span> 0.075<span style="text-decoration: underline;">6</span> 0.075<span style="text-decoration: underline;">6</span> -<span style="color: #BB0000;">0.781</span> -<span style="color: #BB0000;">0.531</span>   1.00    <span style="text-decoration: underline;">4</span>643.    <span style="text-decoration: underline;">3</span>348.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> alpha[5] -<span style="color: #BB0000;">0.444</span> -<span style="color: #BB0000;">0.445</span> 0.084<span style="text-decoration: underline;">2</span> 0.081<span style="text-decoration: underline;">9</span> -<span style="color: #BB0000;">0.580</span> -<span style="color: #BB0000;">0.306</span>   1.00    <span style="text-decoration: underline;">4</span>636.    <span style="text-decoration: underline;">2</span>599.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> alpha[6] -<span style="color: #BB0000;">0.692</span> -<span style="color: #BB0000;">0.692</span> 0.071<span style="text-decoration: underline;">1</span> 0.071<span style="text-decoration: underline;">7</span> -<span style="color: #BB0000;">0.809</span> -<span style="color: #BB0000;">0.576</span>   1.00    <span style="text-decoration: underline;">5</span>498.    <span style="text-decoration: underline;">2</span>871.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: architecture &lt;chr&gt;</span></span></span>
<span></span>
<span><span class="co"># beta estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju</span><span class="op">$</span><span class="va">betas_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 12</span></span></span>
<span><span class="co">#&gt;   variable   mean median    sd   mad    q5    q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> beta[1]  -<span style="color: #BB0000;">1.18</span>  -<span style="color: #BB0000;">1.18</span>  0.266 0.264 -<span style="color: #BB0000;">1.61</span> -<span style="color: #BB0000;">0.739</span>  1.00    <span style="text-decoration: underline;">2</span>019.    <span style="text-decoration: underline;">2</span>899.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> beta[2]  -<span style="color: #BB0000;">1.54</span>  -<span style="color: #BB0000;">1.54</span>  0.261 0.264 -<span style="color: #BB0000;">1.96</span> -<span style="color: #BB0000;">1.10</span>   1.00    <span style="text-decoration: underline;">1</span>930.    <span style="text-decoration: underline;">2</span>502.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> beta[3]  -<span style="color: #BB0000;">0.576</span> -<span style="color: #BB0000;">0.572</span> 0.275 0.280 -<span style="color: #BB0000;">1.03</span> -<span style="color: #BB0000;">0.123</span>  1.00    <span style="text-decoration: underline;">2</span>019.    <span style="text-decoration: underline;">2</span>814.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> beta[4]  -<span style="color: #BB0000;">1.30</span>  -<span style="color: #BB0000;">1.30</span>  0.269 0.263 -<span style="color: #BB0000;">1.75</span> -<span style="color: #BB0000;">0.855</span>  1.00    <span style="text-decoration: underline;">2</span>105.    <span style="text-decoration: underline;">2</span>714.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> beta[5]  -<span style="color: #BB0000;">1.44</span>  -<span style="color: #BB0000;">1.44</span>  0.258 0.259 -<span style="color: #BB0000;">1.85</span> -<span style="color: #BB0000;">1.01</span>   1.01    <span style="text-decoration: underline;">1</span>979.    <span style="text-decoration: underline;">2</span>244.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> beta[6]  -<span style="color: #BB0000;">1.00</span>  -<span style="color: #BB0000;">0.997</span> 0.255 0.255 -<span style="color: #BB0000;">1.43</span> -<span style="color: #BB0000;">0.593</span>  1.00    <span style="text-decoration: underline;">2</span>203.    <span style="text-decoration: underline;">2</span>564.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: is_significant &lt;lgl&gt;, architecture &lt;chr&gt;</span></span></span></code></pre></div>
<p>Note that the estimates are fairly consistent between model type,
even without motif information. Using motif information improves
inference, but is not strictly necessary for keju performance.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>mapping<span class="op">=</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">keju</span><span class="op">$</span><span class="va">alphas_estimate</span><span class="op">$</span><span class="va">mean</span>, y<span class="op">=</span><span class="va">keju_ms</span><span class="op">$</span><span class="va">alphas_estimate</span><span class="op">$</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_abline</span><span class="op">(</span>slope<span class="op">=</span><span class="fl">1</span>, intercept<span class="op">=</span><span class="fl">0</span>, color<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p1</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-1.png" class="r-plt" alt="" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>mapping<span class="op">=</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">keju</span><span class="op">$</span><span class="va">betas_estimate</span><span class="op">$</span><span class="va">mean</span>, y<span class="op">=</span><span class="va">keju_ms</span><span class="op">$</span><span class="va">betas_estimate</span><span class="op">$</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_abline</span><span class="op">(</span>slope<span class="op">=</span><span class="fl">1</span>, intercept<span class="op">=</span><span class="fl">0</span>, color<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-2.png" class="r-plt" alt="" width="700"></p>
<p>For models with motif-level information (motif_shrinkage and
covariate_motif_slope_intercept), we can access motif-level estimates in
the manner shown below.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># transcription rate motif estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju_ms</span><span class="op">$</span><span class="va">alpha_motifs_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 11</span></span></span>
<span><span class="co">#&gt;   variable      mean median     sd    mad     q5     q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> motif_tran… -<span style="color: #BB0000;">0.169</span> -<span style="color: #BB0000;">0.176</span> 0.735  0.696  -<span style="color: #BB0000;">1.36</span>   1.06   1.00     <span style="text-decoration: underline;">4</span>956.    <span style="text-decoration: underline;">3</span>093.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> motif_tran… -<span style="color: #BB0000;">0.518</span> -<span style="color: #BB0000;">0.525</span> 0.308  0.283  -<span style="color: #BB0000;">0.998</span> -<span style="color: #BB0000;">0.014</span><span style="color: #BB0000; text-decoration: underline;">3</span> 1.00     <span style="text-decoration: underline;">3</span>380.    <span style="text-decoration: underline;">2</span>506.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> motif_tran… -<span style="color: #BB0000;">0.731</span> -<span style="color: #BB0000;">0.733</span> 0.090<span style="text-decoration: underline;">7</span> 0.088<span style="text-decoration: underline;">2</span> -<span style="color: #BB0000;">0.878</span> -<span style="color: #BB0000;">0.579</span>  1.000    <span style="text-decoration: underline;">5</span>209.    <span style="text-decoration: underline;">2</span>660.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> motif_tran… -<span style="color: #BB0000;">0.672</span> -<span style="color: #BB0000;">0.672</span> 0.099<span style="text-decoration: underline;">2</span> 0.095<span style="text-decoration: underline;">8</span> -<span style="color: #BB0000;">0.838</span> -<span style="color: #BB0000;">0.507</span>  1.00     <span style="text-decoration: underline;">4</span>353.    <span style="text-decoration: underline;">3</span>019.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> motif_tran…  0.774  0.776 0.208  0.206   0.439  1.11   1.00     <span style="text-decoration: underline;">4</span>720.    <span style="text-decoration: underline;">2</span>813.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> motif_tran…  1.19   1.21  0.311  0.285   0.663  1.69   1.00     <span style="text-decoration: underline;">4</span>029.    <span style="text-decoration: underline;">2</span>407.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: motif &lt;chr&gt;</span></span></span>
<span></span>
<span><span class="co"># effect size motif estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju_ms</span><span class="op">$</span><span class="va">beta_motifs_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 12</span></span></span>
<span><span class="co">#&gt;   variable         mean median    sd   mad     q5    q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> motif_effect[… -<span style="color: #BB0000;">0.547</span> -<span style="color: #BB0000;">0.599</span> 0.820 0.809 -<span style="color: #BB0000;">1.80</span>   0.855  1.00   <span style="text-decoration: underline;">3</span>361.     <span style="text-decoration: underline;">2</span>579.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> motif_effect[… -<span style="color: #BB0000;">0.983</span> -<span style="color: #BB0000;">1.000</span> 0.468 0.452 -<span style="color: #BB0000;">1.73</span>  -<span style="color: #BB0000;">0.189</span>  1.00    213.      538.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> motif_effect[…  0.934  0.930 0.357 0.368  0.353  1.53   1.02     97.6     192.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> motif_effect[…  1.44   1.44  0.362 0.376  0.850  2.04   1.02     97.8     179.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> motif_effect[… -<span style="color: #BB0000;">1.04</span>  -<span style="color: #BB0000;">1.04</span>  0.363 0.381 -<span style="color: #BB0000;">1.62</span>  -<span style="color: #BB0000;">0.440</span>  1.01    101.      188.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> motif_effect[… -<span style="color: #BB0000;">1.21</span>  -<span style="color: #BB0000;">1.21</span>  0.388 0.394 -<span style="color: #BB0000;">1.85</span>  -<span style="color: #BB0000;">0.574</span>  1.01    119.      281.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: is_significant &lt;lgl&gt;, motif &lt;chr&gt;</span></span></span></code></pre></div>
<p>Finally, for covariate_motif_slope_intercept, we can access
covariate-level effects on transcription rate in this manner:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># intercept estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju_cmsi</span><span class="op">$</span><span class="va">covariate_intercept_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 11</span></span></span>
<span><span class="co">#&gt;   variable       mean median    sd   mad     q5    q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> intercept[1]  0.461  0.464 0.538 0.529 -<span style="color: #BB0000;">0.417</span> 1.34    1.01     634.    <span style="text-decoration: underline;">1</span>245.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> intercept[2] -<span style="color: #BB0000;">0.397</span> -<span style="color: #BB0000;">0.402</span> 0.285 0.280 -<span style="color: #BB0000;">0.866</span> 0.061<span style="text-decoration: underline;">6</span>  1.01     742.    <span style="text-decoration: underline;">1</span>434.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> intercept[3] -<span style="color: #BB0000;">0.333</span> -<span style="color: #BB0000;">0.331</span> 0.278 0.266 -<span style="color: #BB0000;">0.791</span> 0.130   1.01     729.    <span style="text-decoration: underline;">1</span>541.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: covariate &lt;chr&gt;</span></span></span>
<span></span>
<span><span class="co"># slope estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">keju_cmsi</span><span class="op">$</span><span class="va">covariate_slope_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 11</span></span></span>
<span><span class="co">#&gt;   variable   mean median    sd   mad      q5   q95  rhat ess_bulk ess_tail</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> slope[1]  0.717  0.659 0.476 0.462  0.039<span style="text-decoration: underline;">5</span> 1.61   1.00    <span style="text-decoration: underline;">1</span>076.    <span style="text-decoration: underline;">1</span>509.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> slope[2] -<span style="color: #BB0000;">0.142</span> -<span style="color: #BB0000;">0.173</span> 0.264 0.259 -<span style="color: #BB0000;">0.514</span>  0.335  1.00    <span style="text-decoration: underline;">1</span>224.    <span style="text-decoration: underline;">1</span>820.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> slope[3] -<span style="color: #BB0000;">0.158</span> -<span style="color: #BB0000;">0.185</span> 0.261 0.252 -<span style="color: #BB0000;">0.528</span>  0.321  1.00    <span style="text-decoration: underline;">1</span>219.    <span style="text-decoration: underline;">2</span>013.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: covariate &lt;chr&gt;</span></span></span></code></pre></div>
<p>Thanks for using keju! Email me at <a href="mailto:albertsxue@gmail.com" class="email">albertsxue@gmail.com</a> with any questions.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Albert Xue.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
